### PR TITLE
Obsolete ServicePointManager 🎉

### LINKED ISSUE #71836: Obsolete ServicePointManager
*Edit:* Changed the proposal to obsolete the entire class.

## Background and motivation

Most of the public API-s exposed by `ServicePointManager` are unused, having no effect other than being copied to `ServicePoint` properties. Even properties with actual effect on legacy API-s could be mistakenly used in combo with `HttpClient`. [Team consensus](https://github.com/dotnet/runtime/issues/62770#issuecomment-993830267) is to obsolete the entire `ServicePointManager` class if possible.

## API proposal

```diff

+ [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
  public class ServicePointManager
  {
-     [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
      public static ServicePoint FindServicePoint(Uri address);

-     [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
      public static ServicePoint FindServicePoint(string uriString, IWebProxy? proxy);

-     [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
      public static ServicePoint FindServicePoint(Uri address, IWebProxy? proxy);
  }
```

## Alternative proposal

Obsolete only unused members.

```csharp
namespace System.Net;

public partial class ServicePointManager
{
    // Remaining members:
    // public static bool CheckCertificateRevocationList { get; set; }
    // public static SecurityProtocolType SecurityProtocol { get; set; }
    // public static RemoteCertificateValidationCallback? ServerCertificateValidationCallback { get; set; }

    [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    public const int DefaultNonPersistentConnectionLimit = 4;
    [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    public const int DefaultPersistentConnectionLimit = 2;
    [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    public static int DefaultConnectionLimit { get; set; }
    [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    public static int DnsRefreshTimeout { get; set; }
    [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    public static bool EnableDnsRoundRobin { get; set; }
    [UnsupportedOSPlatform("browser")]
    [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    public static EncryptionPolicy EncryptionPolicy { get; }
    [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    public static bool Expect100Continue { get; set; }
    [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    public static int MaxServicePointIdleTime { get; set; }
    [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    public static int MaxServicePoints { get; set; }
    [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    public static bool ReusePort { get; set; }
    [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    public static bool UseNagleAlgorithm { get; set; }

    // Already obsolete:
    // [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    // public static ServicePoint FindServicePoint(string uriString, IWebProxy? proxy);
    // [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    // public static ServicePoint FindServicePoint(Uri address);
    // [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    // public static ServicePoint FindServicePoint(Uri address, IWebProxy? proxy);
    // [Obsolete(DiagnosticId = Obsoletions.WebRequestDiagId)]
    // public static void SetTcpKeepAlive(bool enabled, int keepAliveTime, int keepAliveInterval);
}
```

### PR DESCRIPTION
**Pull request by MihaZupan**
Closes #71836

### CODE REVIEW

File: src/libraries/System.Net.HttpListener/src/System/Net/Managed/HttpConnection.cs [hunk 1]
```diff
@@ -100,7 +101,11 @@ public HttpConnection(Socket sock, HttpEndPointListener epl, bool secure, X509Ce
             }
 
             _timer = new Timer(OnTimeout, null, Timeout.Infinite, Timeout.Infinite);
+
+#pragma warning disable SYSLIB0014 // ServicePointManager is obsolete
             _sslStream?.AuthenticateAsServer(_cert, true, (SslProtocols)ServicePointManager.SecurityProtocol, false);
+#pragma warning restore SYSLIB0014
```
Comment (MihaZupan) [inline src/libraries/System.Net.HttpListener/src/System/Net/Managed/HttpConnection.cs:107]: The `SslStream` part of this logic is currently unreachable - tracked by #19752

https://github.com/dotnet/runtime/blob/d7ae8c61f0937df4dcffe88ba96eae65fa523a78/src/libraries/System.Net.HttpListener/src/System/Net/Managed/HttpEndPointListener.cs#L56-L60

https://github.com/dotnet/runtime/blob/d7ae8c61f0937df4dcffe88ba96eae65fa523a78/src/libraries/System.Net.HttpListener/src/System/Net/Managed/HttpListener.Certificates.cs#L20-L24

https://github.com/dotnet/runtime/blob/d7ae8c61f0937df4dcffe88ba96eae65fa523a78/src/libraries/System.Net.HttpListener/src/System/Net/Managed/HttpEndPointListener.cs#L115-L119

File: src/libraries/System.Net.Requests/src/System/Net/FileWebRequest.cs [hunk 2]
```diff
@@ -9,6 +9,8 @@
 
 namespace System.Net
 {
+    // NOTE: While this class is not explicitly marked as obsolete,
+    // it effectively is by virtue of WebRequest.Create being obsolete.
```
Comment (stephentoub) [inline src/libraries/System.Net.Requests/src/System/Net/FileWebRequest.cs:13]: Why haven't we explicitly marked it?

Comment (MihaZupan) [reply src/libraries/System.Net.Requests/src/System/Net/FileWebRequest.cs:13]: I believe it was to reduce the noise for anyone still using it without suppressing it at the project level.
They must suppress the warning at least once when creating the request, but not everywhere else they might have plumbed the types through.

**Review by ManickaP — APPROVED** [review]
:tada:

File: src/libraries/System.Net.Requests/ref/System.Net.Requests.cs [hunk 3]
```diff
@@ -486,6 +486,7 @@ public partial class ServicePoint
         public bool CloseConnectionGroup(string connectionGroupName) { throw null; }
         public void SetTcpKeepAlive(bool enabled, int keepAliveTime, int keepAliveInterval) { }
     }
+    [System.ObsoleteAttribute("WebRequest, HttpWebRequest, ServicePoint, and WebClient are obsolete. Use HttpClient instead.", DiagnosticId = "SYSLIB0014", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
```
Comment (rzikm) [inline src/libraries/System.Net.Requests/ref/System.Net.Requests.cs:489]: I would like it to be mentioned in the message that ServicePointManager members do not affect SslStream and HttpClient behavior (which users mistakenly keep assuming after migrating from .NET Framework).

Comment (MihaZupan) [reply src/libraries/System.Net.Requests/ref/System.Net.Requests.cs]: Added a simple `Settings on ServicePointManager no longer affect SslStream or HttpClient.` suffix to the message (just on ServicePointManager).

**Pull request merged**